### ISSUE TITLE
EVP_DigestUpdate crashes with dgst command-line 

### ISSUE THREAD

**Issue by cpereida** [openssl/openssl#10114]
Hey OpenSSL team, our [Triggerflow](https://gitlab.com/nisec/openssl-triggerflow-ci/pipelines) tool detected a [problem](https://gitlab.com/nisec/openssl-triggerflow-ci/-/jobs/310739852) during some of our testing in master.

Further inspection it seems that `EVP_DigestUpdate` crashes when using the dgst command-line tool to generate DSA signatures. For example,
```
openssl dgst -sha512 -sign dsa.pem -out lsb-release.sig data
```

I believe the issue is in a NULL pointer indirection in `EVP_DigestUpdate`. 
```
legacy:
    return ctx->update(ctx, data, count);
```

### gdb output
```
(gdb) p *ctx
$1 = {reqdigest = 0x7ffff785c520 <sha512_md>, digest = 0x0, engine = 0x0, flags = 0, md_data = 0x0, pctx = 0x55555584c830, update = 0x0, provctx = 0x0, fetched_digest = 0x0}
```

We appreciate if you could take a look into this.

**Comment by mattcaswell** [comment]
This is almost certainly related to #10013 in some way.

**Comment by mattcaswell** [comment]
> I believe the issue is in a NULL pointer indirection in EVP_DigestUpdate.

```
legacy:
    return ctx->update(ctx, data, count);
```

It should not be taking the legacy path in this case.

**Comment by levitte** [comment]
Wasn't something like this fixed already?  Ah yeah, in 5f5c3b4f278bedea466bc4338649b8c540372264...  are there other places that need a similar fix, or does TriggerFlow simply need to catch up?

**Comment by cpereida** [comment]
> Wasn't something like this fixed already? 
I just tested today with the most recent master, and still fails.

**Comment by levitte** [comment]
I pointed at the wrong commit, btw.  The fix I was talking about is in #10082.  However, if we still have a failure...

**Comment by levitte** [comment]
BTW, I see it happen too:

```
: ; ./util/shlib_wrap.sh apps/openssl dgst -sha512 -sign ../master/test/recipes/04-test_pem_data/dsa.pem -out lsb-release.sig ~/.bashrc
Enter pass phrase for ../master/test/recipes/04-test_pem_data/dsa.pem:
Segmenteringsfel
```

A gdb session shows that `*ctx` is void of much provider aware content, this does look genuinely legacy, but still a bit off...

```
(gdb) print *ctx
$1 = {reqdigest = 0x7ffff7e53490 <sha512_md>, digest = 0x0, engine = 0x0, 
  flags = 0, md_data = 0x0, pctx = 0x4e6830, update = 0x0, provctx = 0x0, 
  fetched_digest = 0x0}
(gdb) print *ctx->pctx
$2 = {operation = 64, op = {kex = {exchange = 0x5cadb0, 
      exchprovctx = 0x5ca740}, sig = {signature = 0x5cadb0, 
      sigprovctx = 0x5ca740}}, pmeth = 0x7ffff7edbfc0 <dsa_pkey_meth>, 
  engine = 0x0, pkey = 0x5cacb0, peerkey = 0x0, data = 0x51fbf0, 
  app_data = 0x0, pkey_gencb = 0x0, keygen_info = 0x51fc00, 
  keygen_info_count = 2}
(gdb) print *ctx->pctx->pmeth
$3 = {pkey_id = 116, flags = 2, init = 0x7ffff7c56e90 <pkey_dsa_init>, 
  copy = 0x7ffff7c56f40 <pkey_dsa_copy>, 
  cleanup = 0x7ffff7c56fd0 <pkey_dsa_cleanup>, paramgen_init = 0x0, 
  paramgen = 0x7ffff7c57010 <pkey_dsa_paramgen>, keygen_init = 0x0, 
  keygen = 0x7ffff7c57160 <pkey_dsa_keygen>, sign_init = 0x0, 
  sign = 0x7ffff7c57240 <pkey_dsa_sign>, verify_init = 0x0, 
  verify = 0x7ffff7c57310 <pkey_dsa_verify>, verify_recover_init = 0x0, 
  verify_recover = 0x0, signctx_init = 0x0, signctx = 0x0, 
  verifyctx_init = 0x0, verifyctx = 0x0, encrypt_init = 0x0, encrypt = 0x0, 
  decrypt_init = 0x0, decrypt = 0x0, derive_init = 0x0, derive = 0x0, 
  ctrl = 0x7ffff7c573c0 <pkey_dsa_ctrl>, 
  ctrl_str = 0x7ffff7c57780 <pkey_dsa_ctrl_str>, digestsign = 0x0, 
  digestverify = 0x0, check = 0x0, public_check = 0x0, param_check = 0x0, 
  digest_custom = 0x0}
```

**Comment by mattcaswell** [comment]
Fix in #10116.

**Comment by cpereida** [comment]
> Fix in #10116.

Thanks!
